File: problems/树/94.二叉树的中序遍历.py
#!/usr/bin/env python3
"""
94. 二叉树的中序遍历
"""
from typing import List, Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def inorder_traversal(root: Optional[TreeNode]) -> List[int]:
    result = []
    stack = []
    cur = root
    while cur or stack:
        while cur:
            stack.append(cur)
            cur = cur.left
        cur = stack.pop()
        result.append(cur.val)
        cur = cur.right
    return result

if __name__ == "__main__":
    root1 = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))
    print(inorder_traversal(root1))  # [4, 2, 5, 1, 3]
    print(inorder_traversal(TreeNode(1)))  # [1]
    print(inorder_traversal(None))  # []

File: problems/树/94.二叉树的中序遍历.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<int> inorderTraversal(TreeNode* root) {
    vector<int> result;
    stack<TreeNode*> st;
    TreeNode* cur = root;
    while (cur || !st.empty()) {
        while (cur) { st.push(cur); cur = cur->left; }
        cur = st.top(); st.pop();
        result.push_back(cur->val);
        cur = cur->right;
    }
    return result;
}
int main() {
    TreeNode* root = new TreeNode(1);
    root->left  = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left  = new TreeNode(4);
    root->left->right = new TreeNode(5);
    for (int v : inorderTraversal(root)) cout << v << " "; cout << endl; // 4 2 5 1 3
    for (int v : inorderTraversal(new TreeNode(1))) cout << v << " "; cout << endl; // 1
    for (int v : inorderTraversal(nullptr)) cout << v << " "; cout << endl; // (empty)
    return 0;
}
